src/common/common-thread-private.h

```diff
@@ -123,6 +123,61 @@ int COMMON_PREFIX (thread_create) (bson_thread_t *thread,
 bool COMMON_PREFIX (mutex_is_locked) (bson_mutex_t *mutex);
 #endif
 
+/**
+ * @brief A shared mutex (a read-write lock)
+ *
+ * A shared mutex can be locked in 'shared' mode or 'exclusive' mode. Only one
+ * thread may hold exclusive mode at a time. Any number of threads may hold
+ * the lock in shared mode simultaneously. No thread can hold in exclusive mode
+ * while another thread holds in shared mode, and vice-versa.
+ */
+typedef struct bson_shared_mutex_t {
+   BSON_IF_WINDOWS (SRWLOCK native;)
+   BSON_IF_POSIX (pthread_rwlock_t native;)
```

Comment:
Note: it turns out this change also addresses several C99 conformance warnings currently present on the main branch:
```
src/libbson/src/bson/bson-context.c:255:8: warning: implicit declaration of function 'gethostname' is invalid in C99 [-Wimplicit-function-declaration]
src/libbson/src/bson/bson-iso8601.c:313:7: warning: implicit declaration of function 'gmtime_r' is invalid in C99 [-Wimplicit-function-declaration]
src/libbson/src/bson/bson-context.c:236:13: warning: implicit declaration of function 'rand_r' is invalid in C99 [-Wimplicit-function-declaration]
src/libbson/src/bson/bson-error.c:115:8: warning: implicit declaration of function 'strerror_r' is invalid in C99 [-Wimplicit-function-declaration]
```

Reply:
It looks like setting `_XOPEN_SOURCE=600` hides `strnlen` as well, and it only happens that the ABI is happy with the implicit definitions. Setting `_XOPEN_SOURCE=700` or `_POSIX_C_SOURCE=200109L` both define everything we need. I will add `_XOPEN_SOURCE=700`, as we have no implicit-fns at that level.

Reply:
Rationale behind using `add_compile_definitions()` instead of `target_compile_definitions()` as suggested earlier (motivated by [modern CMake practices](https://gist.github.com/mbinna/c61dbb39bca0e4fb7d1f73b0d66a4fd1#targets-and-properties))?

Reply:
It's somewhat arbitrary in this case, but directory-attached properties are good for cases where you want to compile the sources within a directory with certain settings, versus attaching them as private options on the sources of particular targets. Because there are multiple targets compiling the same sources, and we want all sources in the directory to see those options, it felt more sensible to attach the option to the directory instead.